packages/graph/hash_graph/lib/graph/src/store/query/mod.rs

```diff
@@ -147,106 +167,177 @@ impl Expression {
+        fmt.write_str("evaluation of expression failed")
+    }
+}
+
+impl Error for ExpressionError {}
+
 impl Expression {
-    // TODO: Implement error handling
-    //   see https://app.asana.com/0/0/1202884883200968/f
-    #[expect(clippy::missing_panics_doc, reason = "Error handling not applied yet")]
-    pub fn evaluate<'a, R, C>(&'a self, resolver: &'a R, context: &'a C) -> BoxFuture<Literal>
+    pub fn evaluate<'a, R, C>(
+        &'a self,
+        resolver: &'a R,
+        context: &'a C,
+    ) -> BoxFuture<Result<Literal, ExpressionError>>
     where
         R: Resolve<C> + Sync,
         C: Sync,
     {
         async move {
-            match self {
+            Ok(match self {
                 Expression::Eq(expressions) => {
                     for expression in expressions.windows(2) {
                         if !compare(
-                            &expression[0].evaluate(resolver, context).await,
-                            &expression[1].evaluate(resolver, context).await,
-                        ) {
-                            return Literal::Bool(false);
+                            &expression[0].evaluate(resolver, context).await?,
+                            &expression[1].evaluate(resolver, context).await?,
+                        )? {
+                            return Ok(Literal::Bool(false));
                         }
                     }
                     Literal::Bool(true)
                 }
                 Expression::Ne(expressions) => {
                     for expression in expressions.windows(2) {
                         if compare(
-                            &expression[0].evaluate(resolver, context).await,
-                            &expression[1].evaluate(resolver, context).await,
-                        ) {
-                            return Literal::Bool(false);
+                            &expression[0].evaluate(resolver, context).await?,
+                            &expression[1].evaluate(resolver, context).await?,
+                        )? {
+                            return Ok(Literal::Bool(false));
                         }
                     }
                     Literal::Bool(true)
                 }
                 Expression::All(expressions) => {
                     for expression in expressions {
-                        match expression.evaluate(resolver, context).await {
+                        match expression.evaluate(resolver, context).await? {
                             Literal::Bool(true) => continue,
-                            literal @ Literal::Bool(false) => return literal,
-                            literal => panic!("Not a boolean: {literal:?}"),
+                            literal @ Literal::Bool(false) => return Ok(literal),
+                            literal => bail!(
+                                Report::new(ExpressionError)
+                                    .attach_printable(format!("Not a boolean: {literal:?}"))
```

Comment:
Can you give me an example of how this can occur?

Reply:
As a sidenote, do we want capitalised attachments? Or should we follow convention for error messages?

Reply:
This could happen if you pass in an `Expression` which is not a boolean, e.g.
```http
GET http://127.0.0.1:4000/property-types
Content-Type: application/json

{
  "literal": 10
}
```

This will result in
```
   0: evaluation of expression failed
             at lib/graph/src/store/postgres/ontology/property_type/mod.rs:125:25
      - does not result in a boolean value
```

Reply:
We should follow the conventions for error. I applied the changes in 290f8b692ff7d939853438291c1c38ca98ef422c and opened https://github.com/hashintel/hash/issues/1017 for that.

Reply:
Thanks